### ISSUE TITLE
Provide a way to implement an infinite / streaming collection as a Dataset

### ISSUE THREAD

**Issue by concretevitamin** [pytorch/pytorch#17451]
I’d like to implement an infinite loop Dataset & DataLoader. Here’s what I tried:
```python
class Infinite(Dataset):
    def __len__(self):
        return HPARAMS.batch_size
#         return 1<<30 # This causes huge memory usage.
    def __getitem__(self, idx):
        """Randomly generates one new example."""
        return sample_func_to_be_parallelized()

infinite_loader = DataLoader(
    dataset=Infinite(), 
    batch_size=HPARAMS.batch_size, 
    num_workers=16,
    worker_init_fn=lambda worker_id: np.random.seed(worker_id),  
)

while True:
    for idx, data in enumerate(infinite_loader):
        # forward + backward on "data"
```
As you can see, the main challenge here is the `__len()__` method. If I put a large enough number there, like 1<<30, the symptom is memory usage will JUMP TO 10+GB on the first iteration of train loop. After a while the workers are killed presumably due to OOM.

If I put a small number there, like 1, I believe there's a huge operational overhead and I did not observe speedups.

I’m guessing the culprit of the excessive memory usage is somewhere in the stack, a bunch of things are cached. Upon a casual look at the Python side of things I can’t pinpoint where.  (Possibly `np.permutation(large_val)`.)

**It'd be great to have a way to implement such infinite collection of examples.  Another interpretation is it streams in new data infinitely.**

**Comment by soumith** [comment]
`return HPARAMS.batch_size` looks pretty wrong. you would want some large value (as you mentioned, something like `1<<30`).
The DataLoader only pre-loads one batch ahead of time.
What you are running into is probably that if the size is VERY large, we actually generate indices ahead of time to preserve determinism: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L144-L148

Also, the DataLoader is not a one-stop solution for all situations, something like `joblib` might serve better here. @SsnL has a WIP PR for IterableDataset, which might also serve this better: https://github.com/pytorch/pytorch/pull/14705

**Comment by concretevitamin** [comment]
Shouldn't it be possible to make a simple optimization: if `shuffle` is set to False, don't pre-generate the indices.
